src/main/scala/clump/ClumpSource.scala

```diff
@@ -16,14 +15,30 @@ class ClumpSource[T, U](val fetch: Set[T] => Future[Map[T, U]], val maxBatchSize
 
   def get(input: T): Clump[U] =
     new ClumpFetch(input, ClumpContext().fetcherFor(this))
+
+  def maxBatchSize(int: Int) =
+    new ClumpSource(fetch, maxBatchSize = int)
 }
 
 object ClumpSource {
-  def zip[T, U](fetch: List[T] => Future[List[U]], maxBatchSize: Int): ClumpSource[T, U] = {
+
+  def apply[T, U, C](fetch: C => Future[Iterable[U]])(keyExtractor: U => T)(implicit cbf: CanBuildFrom[Nothing, T, C]) =
+    new ClumpSource(extractKeys(adaptInput(fetch), keyExtractor))
+
+  def from[T, U, C](fetch: C => Future[Map[T, U]])(implicit cbf: CanBuildFrom[Nothing, T, C]) =
+    new ClumpSource(adaptInput(fetch))
+
+  def zip[T, U](fetch: List[T] => Future[List[U]]): ClumpSource[T, U] = {
```

Comment:
Just a stylistic thing but you could match `apply` and `from` by extracting the zip logic like:
```scala
  def zip[T, U](fetch: List[T] => Future[List[U]]): ClumpSource[T, U] = {
    new ClumpSource(zipped(fetch))
  }
  
  private def zipped[T, U](fetch: List[T] => Future[List[U]]) = {
    val zip: List[T] => Future[Map[T, U]] = { inputs =>
      fetch(inputs).map(inputs.zip(_).toMap)
    }
    val setToList: Set[T] => List[T] = _.toList
    setToList.andThen(zip)
  }
```